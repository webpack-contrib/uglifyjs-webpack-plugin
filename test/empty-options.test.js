import { RawSource } from 'webpack-sources';
import UglifyJsPlugin from '../src/index';
import {
  PluginEnvironment,
  cleanErrorStack,
  createCompiler,
  compile,
} from './helpers';

describe('when applied with no options', () => {
  let eventBindings;
  let eventBinding;

  beforeEach(() => {
    const pluginEnvironment = new PluginEnvironment();
    const compilerEnv = pluginEnvironment.getEnvironmentStub();
    compilerEnv.context = '';

    const plugin = new UglifyJsPlugin();
    plugin.apply(compilerEnv);
    eventBindings = pluginEnvironment.getEventBindings();
  });

  it('matches snapshot', () => {
    const compiler = createCompiler();
    new UglifyJsPlugin().apply(compiler);


    return compile(compiler).then((stats) => {
      const errors = stats.compilation.errors.map(cleanErrorStack);
      const warnings = stats.compilation.warnings.map(cleanErrorStack);

      expect(errors).toMatchSnapshot('errors');
      expect(warnings).toMatchSnapshot('warnings');

      for (const file in stats.compilation.assets) {
        if (Object.prototype.hasOwnProperty.call(stats.compilation.assets, file)) {
          expect(stats.compilation.assets[file].source()).toMatchSnapshot(file);
        }
      }
    });
  });

  it('binds one event handler', () => {
    expect(eventBindings.length).toBe(1);
  });

  describe('compilation handler', () => {
    beforeEach(() => {
      [eventBinding] = eventBindings;
    });

    it('binds to compilation event', () => {
      expect(eventBinding.name).toBe('compilation');
    });

    describe('when called', () => {
      let chunkPluginEnvironment;
      let compilationEventBindings;
      let compilationEventBinding;
      let compilation;
      let callback;

      beforeEach(() => {
        chunkPluginEnvironment = new PluginEnvironment();
        compilation = chunkPluginEnvironment.getEnvironmentStub();
        compilation.assets = {
          'test.js': {},
          'test1.js': '',
          'test2.js': {
            source: () => 'invalid javascript',
          },
          'test3.js': {
            source: () => '/** @preserve Foo Bar */ function foo(longVariableName) { longVariableName = 1; }',
          },
        };
        compilation.errors = [];

        eventBinding.handler(compilation);
        compilationEventBindings = chunkPluginEnvironment.getEventBindings();
      });

      it('binds one event handler', () => {
        expect(compilationEventBindings.length).toBe(1);
      });

      describe('optimize-chunk-assets handler', () => {
        beforeEach(() => {
          [compilationEventBinding] = compilationEventBindings;
        });

        it('binds to optimize-chunk-assets event', () => {
          expect(compilationEventBinding.name).toEqual('optimize-chunk-assets');
        });

        it('only calls callback once', () => {
          callback = jest.fn();
          compilationEventBinding.handler([''], callback);
          expect(callback.mock.calls.length).toBe(1);
        });

        it('default only parses filenames ending with .js', () => {
          compilationEventBinding.handler([{
            files: ['test', 'test.js'],
          }], () => {
            expect(Object.keys(compilation.assets).length).toBe(4);
          });
        });

        it('early returns if private property is already set', () => {
          compilationEventBinding.handler([{
            files: ['test.js'],
          }], () => {
            expect(compilation.assets['test.js']).toEqual({});
          });
        });

        it('outputs stack trace errors for invalid asset', () => {
          compilationEventBinding.handler([{
            files: ['test1.js'],
          }], () => {
            expect(compilation.errors.length).toBe(1);
            expect(compilation.errors[0]).toBeInstanceOf(Error);
            expect(compilation.errors[0].message).toEqual(expect.stringContaining('TypeError'));
          });
        });

        it('outputs parsing errors for invalid javascript', () => {
          compilationEventBinding.handler([{
            files: ['test2.js'],
          }], () => {
            expect(compilation.errors.length).toBe(1);
            expect(compilation.errors[0]).toBeInstanceOf(Error);
            expect(compilation.errors[0].message).toEqual(expect.stringContaining('Unexpected token'));
            expect(compilation.errors[0].message).toEqual(expect.stringContaining('[test2.js:1,8]'));
          });
        });

        it('outputs no errors for valid javascript', () => {
          compilationEventBinding.handler([{
            files: ['test3.js'],
          }], () => {
            expect(compilation.errors.length).toBe(0);
          });
        });

        it('outputs RawSource for valid javascript', () => {
          compilationEventBinding.handler([{
            files: ['test3.js'],
          }], () => {
            expect(compilation.assets['test3.js']).toBeInstanceOf(RawSource);
          });
        });

        it('outputs mangled javascript', () => {
          compilationEventBinding.handler([{
            files: ['test3.js'],
          }], () => {
            // eslint-disable-next-line no-underscore-dangle
            expect(compilation.assets['test3.js']._value)
              .not.toEqual(expect.stringContaining('longVariableName'));
          });
        });

        it('compresses and does not output beautified javascript', () => {
          compilationEventBinding.handler([{
            files: ['test3.js'],
          }], () => {
            // eslint-disable-next-line no-underscore-dangle
            expect(compilation.assets['test3.js']._value).not.toEqual(expect.stringContaining('\n'));
          });
        });

        it('preserves comments', () => {
          compilationEventBinding.handler([{
            files: ['test3.js'],
          }], () => {
            // eslint-disable-next-line no-underscore-dangle
            expect(compilation.assets['test3.js']._value).toEqual(expect.stringContaining('/**'));
          });
        });
      });
    });
  });
});
