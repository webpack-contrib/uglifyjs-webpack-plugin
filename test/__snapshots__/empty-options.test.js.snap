// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors 1`] = `Array []`;

exports[`main.6c966be0e6148d71540d.js 1`] = `"(window.webpackJsonp=window.webpackJsonp||[]).push([[\\"main\\"],{\\"./test/fixtures/entry.js\\":function(module,exports){eval(\\"// foo\\\\n/* @preserve*/\\\\n// bar\\\\nconst a = 2 + 2;\\\\n\\\\nmodule.exports = function Foo() {\\\\n  const b = 2 + 2;\\\\n  console.log(b + 1 + 2);\\\\n};\\\\n\\\\n\\\\n//////////////////\\\\n// WEBPACK FOOTER\\\\n// ./test/fixtures/entry.js\\\\n// module id = ./test/fixtures/entry.js\\\\n// module chunks = main\\\\n\\\\n//# sourceURL=webpack:///./test/fixtures/entry.js?\\")}},[[\\"./test/fixtures/entry.js\\",\\"manifest\\",\\"main\\"]]]);"`;

exports[`manifest.ae205c79cb278d0fd374.js 1`] = `"!function(e){function r(r){for(var n,a,c,s=r[0],l=r[1],p=r[2],f=0,d=[];f<s.length;f++)a=s[f],o[a]&&d.push(o[a][0]),o[a]=0;for(n in l)Object.prototype.hasOwnProperty.call(l,n)&&(e[n]=l[n]);for(u&&u(r);d.length;)d.shift()();for(i.push.apply(i,p||[]),f=0;f<i.length;f++){for(var h=i[f],m=!0,v=1;v<h.length;v++){var g=h[v];0!==o[g]&&(m=!1)}m&&(i.splice(f--,1),c=t(t.s=h[0]))}return c}function t(r){if(n[r])return n[r].exports;var o=n[r]={i:r,l:!1,exports:{}};return e[r].call(o.exports,o,o.exports,t),o.l=!0,o.exports}var n={},o={manifest:0},i=[];t.e=function(e){var r=[],n=o[e];if(0!==n)if(n)r.push(n[2]);else{var i=new Promise(function(r,t){n=o[e]=[r,t]});r.push(n[2]=i);var a=document.getElementsByTagName(\\"head\\")[0],u=document.createElement(\\"script\\");u.charset=\\"utf-8\\",u.timeout=12e4,t.nc&&u.setAttribute(\\"nonce\\",t.nc),u.src=t.p+\\"\\"+e+\\".\\"+({main:\\"main\\"}[e]||e)+\\".\\"+{main:\\"6c966be0e6148d71540d\\"}[e]+\\".js\\";var c=setTimeout(function(){s({type:\\"timeout\\",target:u})},12e4);u.onerror=u.onload=s;function s(r){u.onerror=u.onload=null,clearTimeout(c);var t=o[e];if(0!==t){if(t){var n=r&&(\\"load\\"===r.type?\\"missing\\":r.type),i=r&&r.target&&r.target.src,a=new Error(\\"Loading chunk \\"+e+\\" failed.\\\\n(\\"+n+\\": \\"+i+\\")\\");a.type=n,a.request=i,t[1](a)}o[e]=void 0}}a.appendChild(u)}return Promise.all(r)},t.m=e,t.c=n,t.d=function(e,r,n){t.o(e,r)||Object.defineProperty(e,r,{configurable:!1,enumerable:!0,get:n})},t.r=function(e){Object.defineProperty(e,\\"__esModule\\",{value:!0})},t.n=function(e){var r=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(r,\\"a\\",r),r},t.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)},t.p=\\"\\",t.oe=function(e){throw console.error(e),e};var a=window.webpackJsonp=window.webpackJsonp||[],u=a.push.bind(a);a.push=r,a=a.slice();for(var c=0;c<a.length;c++)r(a[c])}([]);"`;

exports[`warnings 1`] = `Array []`;
